packages/kbn-i18n/src/angular/provider.js

```diff
@@ -0,0 +1,62 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { i18n } from '../i18n';
+
+export function i18nProvider() {
+  this.addMessages = function(messages, locale) {
```

Comment:
For all of these methods can't we simply do this?
```js
  this.addMessages = i18n.addMessages;
  this.getMessages = i18n.getMessages;
```
Looking at the implementation of `translate`, I don't _think_ `$get` is actually that different. Internally all the `this.[whatever]` are used to grab values from this provider, which is a simple wrapper around the `i18n` instance. Does the following not work?
```js
  this.$get = i18n.translate;
```

Reply:
Okay, I have some thoughts now that I understand what we're doing.
Why I think there is something weird/antipattern/code smell:

* Normally when we use JavaScript `class` syntax, if the class's methods use `this` internally, we design it that way so that any instance of that `class` can pass along its own `this` context, and the methods use that context. However in this case, we always want to use the `i18n` instance's `this`. 
* The angular provider (`i18nProvider`) is not an instance or a class, yet it is setting its own `this.[properties]`. That's a strange thing. Is it because of how angular expects the provider? Forgive me I don't remember exactly how angular 1.x works and maybe this is just something we have to do.
* Additionally, the capital-case `I18n` is used as a singleton (as `i18n`), so the `this` is always going to be the same, right? 
* And also, arrow functions are normally good for solving issues of the wrong `this` context, but using them here would be absolutely wrong. It would pass along the wrong `this`. That seems to me like we're creating unnecessary indirection within `i18n`. 
* It’s kind of weird to have to do `i18n.translate.bind(i18n)` because it seems like calling translate on the `i18n` should already provide the right `this` context. We’re calling it on the i18n object, after all.

All this points to how `I18n` is designed. Would it work to forgo the class, make all the methods no longer call `this.[whatever]`, and make I18n a function that accepts messages/locale and return an object that do its methods on that? It would follow that the methods of `I18n` would not have to explicitly call `this.[whatever]` because there's no need to take the `this` context--it's always the same context. And thus the angular provider would be something more like a simple wrapper around `I18n`. Basically, if you want to talk about this in the classical way, this would become a fully static class. Does that make sense, and would it make sense for this library?

Something like this?
```js
export function I18n(messages, ...) {
  const _messages = // set it
  const _currentLocale = // set it

  return {
    getMessages: function() {
      return _messages[_currentLocale];
    },
    // ...
  };
}
```
There would be less indirection, less potential confusion around `this` because it isn't actually being used as intended, and I think the angular provider becomes simpler, too. What do you think?

Reply:
I removed I18n class and made i18n a singleton because I don't see real use cases for multiple I18n instances. So we no longer use `this` keyword in i18n engine.